test/MediatR.Tests/GenericRequestHandlerTests.cs

```diff
@@ -0,0 +1,190 @@
+
+            int expectedCombinations = CalculateTotalCombinations(numberOfClasses, numberOfInterfaces, numberOfTypeParameters);
+
+            var testClasses = Enumerable.Range(1, numberOfClasses)
+                .Select(i => dynamicAssembly.GetType($"TestClass{i}")!)
+                .ToArray();
+
+            var combinations = GenerateCombinations(testClasses, numberOfInterfaces);          
+
+            foreach (var combination in combinations)
+            {
+                var concreteRequestType = dynamicRequestType.MakeGenericType(combination);
+                var requestHandlerInterface = typeof(IRequestHandler<>).MakeGenericType(concreteRequestType);
+
+                var handler = provider.GetService(requestHandlerInterface);
+                handler.ShouldNotBeNull($"Handler for {concreteRequestType} should not be null");
+            }            
+        }
+
+        [Theory]
+        [InlineData(9, 3, 3)]
+        [InlineData(10, 4, 4)]
+        [InlineData(1, 1, 1)]
+        [InlineData(50, 3, 3)]
+        public void ShouldRegisterTheCorrectAmountOfHandlers(int numberOfClasses, int numberOfInterfaces, int numberOfTypeParameters)
+        {  
+            var dynamicAssembly = GenerateCombinationsTestAssembly(numberOfClasses, numberOfInterfaces, numberOfTypeParameters);          
+            int expectedCombinations = CalculateTotalCombinations(numberOfClasses, numberOfInterfaces, numberOfTypeParameters);
+            var testClasses = Enumerable.Range(1, numberOfClasses)
+               .Select(i => dynamicAssembly.GetType($"TestClass{i}")!)
+               .ToArray();
+            var combinations = GenerateCombinations(testClasses, numberOfInterfaces);
+            combinations.Count.ShouldBe(expectedCombinations, $"Should have tested all {expectedCombinations} combinations");
+        }
+
+        [Theory]
+        [InlineData(9, 3, 3)]
+        [InlineData(10, 4, 4)]
+        [InlineData(1, 1, 1)]
+        [InlineData(50, 3, 3)]
+        public void ShouldNotRegisterDuplicateHandlers(int numberOfClasses, int numberOfInterfaces, int numberOfTypeParameters)
+        {
+            var dynamicAssembly = GenerateCombinationsTestAssembly(numberOfClasses, numberOfInterfaces, numberOfTypeParameters);
+            int expectedCombinations = CalculateTotalCombinations(numberOfClasses, numberOfInterfaces, numberOfTypeParameters);
+            var testClasses = Enumerable.Range(1, numberOfClasses)
+               .Select(i => dynamicAssembly.GetType($"TestClass{i}")!)
+               .ToArray();
+            var combinations = GenerateCombinations(testClasses, numberOfInterfaces);
+            var hasDuplicates = combinations
+              .Select(x => string.Join(", ", x.Select(y => y.Name)))
+              .GroupBy(x => x)
+              .Any(g => g.Count() > 1);
+
+            hasDuplicates.ShouldBeFalse();
+        }
+
+        [Fact]
+        public void ShouldThrowExceptionWhenRegisterningHandlersWithNoConstraints()

```

Comment:
Why do you need to enforce this?

Reply:
because if no constraint is present then the the registration will try to register all types from the assembly when it isn't necessary. Type constraints should be present to limit the types that can close this generic type.

There isn't any way of knowing how many types are referenced by an assembly and MediatR will try and use to close the request handler types.

Reply:
If the consensus is that this shouldn't be enforced then I'm ok with that. Just say the word and I'll change it.

Reply:
> the registration will try to register all types from the assembly when it isn't necessary

Do I take it then that if you have `MyGenericRequest<T> where T : class` and `MyGenericRequestHandler<T>` then this will register an instance of `MyGenericRequestHandler<T>` for every class in the assembly?  And what happens if the type T is not in the same assembly as `MyGenericRequestHandler`?

Reply:
I'm not sure if I understand your question.. I don't know what you mean by "register an instance". As far as I understand, registering a service is nothing more than an entry in a dictionary, more or less, of a keyed "service type" and a value "implementation type".  no instances are created unless you call `GetService<TService>()` which is what gets called behind the scenes most likely when you stuff a service in a constructor in .net core.

So no, an instance of `MyGenericRequestHandler<T>` is not created for each class in the assemblies.  But a registration entry is registered in the service collection for every single type that closes `MyGenericRequestHandler<T>`..  It is doing the exact same thing you are doing by manually creating closed registrations for your generic handlers.  It is just doing it by scanning the assembly and figuring out which classes close that type. and then registering them. In my opinion if you have have to manually create concrete handlers for each and every generic implementation, then it kind of defeats the purpose. 

Which is why I believe that checking for type constraints is, in my opinion, necessary.

Reply:
Sorry I misspoke, I don't mean instances.  

My point is that type constraints might result in too many types being registered.  Probably worth making it clearer in the wiki how this works with type constraints so people can define suitably specific constraints.

Also to my other question how does the reflection handle looking up types?

So if I had something like this:

```
Domain.csproj
- MyGenericRequest<TEntity, TDto>
- MyGenericHandler<TEntity, TDto>
- MyType/MyTypeDto

Entities.csproj
- MyType
```

```cs
services.AddMediatR(cfg => 
{
    cfg.RegisterHandlersFromAssemblyContaining<MyTypeDto>();
});

services.GetInstance<IMediator>().Send(new MyGenericRequest<MyType, MyTypeDto>());
```

Will this work?

Reply:
The code in the pr is set to throw an exception if you have a generic request handler that you are trying to register that 1) has more than two type parameters **AND** 2) does not have a type constraint defined.  This is to prevent too many registrations from happening.

from: ServiceRegistrar.cs

```csharp
 if (constraintsForEachParameter.Count > 2 && constraintsForEachParameter.Any(constraints => !constraints.Where(x => x.IsInterface || x.IsClass).Any()))
     throw new ArgumentException($"Error registering the generic handler type: {openRequestHandlerImplementation.FullName}. When registering generic requests with more than two type parameters, each type parameter must have at least one constraint of type interface or class.");
```

So, yes that example will work.

Reply:
If you intend on being this critical then I suggest that you clone the repository and test it yourself. Then you can be sure what works and what doesn't.  I'm not really trying to help you make your specific solution work as much as just fixing the issues mentioned in the PR.

Reply:
> Domain.csproj
> - MyGenericRequest<TEntity, TDto>
> - MyGenericHandler<TEntity, TDto>
> - MyType/MyTypeDto
> 
> Entities.csproj
> - MyType
> ```
> 
> ```cs
> services.AddMediatR(cfg => 
> {
>     cfg.RegisterHandlersFromAssemblyContaining<MyTypeDto>();
> });
> 
> services.GetInstance<IMediator>().Send(new MyGenericRequest<MyType, MyTypeDto>());
> ```
> 
> Will this work?

To make this work you would simply need to make sure that the assemblies are added to the `AddMediatR` call.
```csharp
builder.Services.AddMediatR(cfg => cfg.RegisterServicesFromAssemblies([typeof(MyTypeDto).Assembly, typeof(MyType).Assembly]);
```

Reply:
> Sorry I misspoke, I don't mean instances.
> 
> My point is that type constraints might result in too many types being registered. Probably worth making it clearer in the wiki how this works with type constraints so people can define suitably specific constraints.
> 

I don't understand how type constraints could do anything except reduce the number of services being registered. And in most cases explicitly target them.

Reply:
I'm going to resolve this.. If you feel strongly that there should not be enforcement of type constraints on generic handlers with three or more type parameters then please unresolve this or start a new conversation.